lib/device-log/ios-simulator-log.js

```diff
@@ -0,0 +1,119 @@
+import _ from 'lodash';
+import { IOSLog } from './ios-log';
+import { logger } from 'appium-support';
+import { SubProcess, exec } from 'teen_process';
+
+const log = logger.getLogger('IOSSimulatorLog');
+
+const START_TIMEOUT = 10000;
+
+class IOSSimulatorLog extends IOSLog {
+  constructor (opts) {
+    super();
+    this.sim = opts.sim;
+    this.showLogs = !!opts.showLogs;
+    this.xcodeVersion = opts.xcodeVersion;
+    this.proc = null;
+  }
+
+  async startCapture () {
+    if (_.isUndefined(this.sim.udid)) {
+      throw new Error(`Log capture requires a sim udid`);
+    }
+
+    if (!await this.sim.isRunning()) {
+      throw new Error(`iOS Simulator with udid ${this.sim.udid} is not running`);
+    }
+    const tool = 'xcrun';
+    const args = ['simctl', 'spawn', this.sim.udid, 'log', 'stream', '--style', 'compact'];
+    log.debug(`Starting log capture for iOS Simulator with udid '${this.sim.udid}', ` +
+                `using '${tool} ${args.join(' ')}'`);
+    try {
+      // cleanup existing listeners if the previous session has not been terminated properly
+      await exec('pkill', ['-xf', [tool, ...args].join(' ')]);
+    } catch (ign) {}
+    try {
+      this.proc = new SubProcess(tool, args);
+      await this.finishStartingLogCapture();
+    } catch (e) {
+      throw new Error(`Simulator log capture failed. Original error: ${e.message}`);
+    }
+  }
+
+  async finishStartingLogCapture () {
+    if (!this.proc) {
+      log.errorAndThrow('Could not capture device log');
+    }
+    let firstLine = true;
+    this.proc.on('output', (stdout, stderr) => {
+      if (stdout) {
+        if (firstLine) {
+          if (stdout.substr(-1, 1) === '\n') {
+            // don't store the first line of the log because it came before the sim or device was launched
+            firstLine = false;
+          }
+        } else {
+          this.logRow += stdout;
```

Comment:
there might be a potential memory leak

Reply:
This should be fine. The code already runs in `appium-ios-driver` for years. I wan't to clean it up later